pkg/controller/sync/controller.go

```diff
@@ -377,3 +367,145 @@ func (s *FederationSyncController) syncToClusters(fedResource FederatedResource,
 
 	return status
 }
+
+func (s *FederationSyncController) ensureDeletion(fedResource FederatedResource) error {
+	fedResource.DeleteVersions()
+
+	key := fedResource.FederatedName().String()
+	kind := fedResource.FederatedKind()
+
+	glog.V(2).Infof("Ensuring deletion of %s %q", kind, key)
+
+	obj := fedResource.Object()
+
+	finalizers := sets.NewString(obj.GetFinalizers()...)
+	if !finalizers.Has(FinalizerSyncController) {
+		glog.V(2).Infof("%s %q does not have the %q finalizer. Nothing to do.", kind, key, FinalizerSyncController)
+		return nil
+	}
+
+	annotations := obj.GetAnnotations()
+	orphanResources := annotations != nil && annotations[OrphanManagedResources] == "true"
+	if orphanResources {
+		glog.V(2).Infof("Found %q annotation on %s %q. Removing the finalizer.", OrphanManagedResources, kind, key)
+		err := s.removeFinalizer(fedResource)
+		if err != nil {
+			wrappedErr := errors.Wrapf(err, "failed to remove finalizer %q from %s %q", OrphanManagedResources, kind, key)
+			runtime.HandleError(wrappedErr)
+			return wrappedErr
+		}
+		return nil
+		// TODO(marun) Implement removal of labels
+	}
+
+	glog.V(2).Infof("Deleting resources managed by %s %q from member clusters.", kind, key)
+	err := s.deleteFromClusters(fedResource)
+	if err != nil {
+		wrappedErr := errors.Wrapf(err, "failed to delete %s %q", kind, key)
+		runtime.HandleError(wrappedErr)
+		return wrappedErr
+	}
+	return nil
+}
+
+func (s *FederationSyncController) deleteFromClusters(fedResource FederatedResource) error {
+	kind := fedResource.TargetKind()
+	qualifiedName := fedResource.TargetName()
+
+	remainingClusters := []string{}
+	ok, err := s.handleDeletionInClusters(kind, qualifiedName, func(dispatcher dispatch.UnmanagedDispatcher, clusterName string, clusterObj *unstructured.Unstructured) {
+		if fedResource.IsNamespaceInHostCluster(clusterObj) {
+			return
+		}
+
+		remainingClusters = append(remainingClusters, clusterName)
+
+		if clusterObj.GetDeletionTimestamp() != nil {
+			return
+		}
+
+		dispatcher.Delete(clusterName)
+	})
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return errors.Errorf("failed to remove managed resources from one or more clusters.")
+	}
+	if len(remainingClusters) > 0 {
+		fedKind := fedResource.FederatedKind()
+		fedName := fedResource.FederatedName()
+		glog.V(2).Infof("Waiting for resources managed by %s %q to be removed from the following clusters: %s", fedKind, fedName, strings.Join(remainingClusters, ", "))
+		return nil
+	}
+	// Managed resources no longer exist in any member cluster
+	return s.removeFinalizer(fedResource)
+}
```

Comment:
I am thinking is there a possibility that reconciliation is never triggered for the resource as we return nil and keep waiting for next reconciliation. The finalizer on the resource stays and federated resource deletion never runs to completion.

Reply:
The only nil return in this function is from line 439, and that path is only followed if one more more operations was executed successfully against one or more member clusters.  I've been assuming that ensures that one or more reconcile-triggering events would be pending, but is that not correct?

Reply:
Correct me if I'm wrong, In case of deletion, the reconcile may not be triggered as now in federated informer we do watch based on `managed` label.